keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    You can find a snippet of validation rules below:
+
+    ```yaml
+    claimValidationRules:
+    
+    - rule: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
+    ```
+
+3. `ClaimMappings` - rules to map claims from a token to Kubernetes user attributes.
+    ```go
+    type UserAttributes struct {
+        // Username represents an option for the username attribute.
+        Username string
+        // Groups represents an option for the groups attribute. 
+        // +optional
+        Groups string
+        // UID represents an option for the uid attribute.
+        // +optional
+        UID string
+        // Extra represents an option for the extra attribute.
+        // +optional
+        Extra []ExtraMapping
+    }
+   
+    type ExtraMapping struct {
+        // Key is a CEL expression to extract extra attribute key.
+        Key string
+        // Expression is a CEL expression to extract extra attribute value.
+        Expression string
+    }
+    ```
+
+    Every field of structures above is a CEL expression. For username and uid, the expression should evaluate a string.
+    For groups, it should evaluate to an array. A map of extra attributes is composed of key/value pairs, each with their
+    own CEL expressions, both should evaluate to a string.
+
+    The example of mapping user attributes can be found below:
+
+    ```yaml
+    claimMappings:
+      username: 'claims.username + ":external-user"'
+      groups: 'claims.roles.split(",")'
+      uid: 'claims.sub'
+      extra:
+      - key: '"client_name"'
+        expression: 'claims.aud'
+    ```
+   
+    For the token with the following claims
+    ```json
+    {
+      "sub": "119abc",
+      "aud": "kubernetes",
+      "username": "jane_doe",
+      "roles": "admin,user",
+      ...
+    }
+    ```
+    our authenticator will extract the following user attributes:
+    ```yaml
+    username: jane_doe:external-user
+    groups: ["admin", "user"]
+    uid: "119abc"
+    extra:
+      client_name: kubernetes
+    ```
+4. `ClaimsFilter` - list of claim names that should be passed to CEL expressions. The assumption is that administrators
+   know the structure of the token and the exact claims they will use in CEL expressions.
+   This option helps to reduce system load and operate only with required claims.
+
+### More about CEL
+
+* CEL runtime should be compiled only once if structured OIDC config option is enabled
+* To make working with strings more convinient, `strings` and `encoding` [CEL extensions](https://github.com/google/cel-go/tree/v0.9.0/ext) should be enabled, 
```

Comment:
This can be generalized to something like "The standard Kubernetes CEL environment, including extension libraries, will be used".

Here's the current environment: https://github.com/kubernetes/kubernetes/blob/5fe3563ad7e04d5470368aa821f42f131d3bd8fc/staging/src/k8s.io/apiserver/pkg/cel/library/libraries.go#L26, https://github.com/kubernetes/kubernetes/blob/5fe3563ad7e04d5470368aa821f42f131d3bd8fc/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L83

We will need to add the `encoding` library to the environment since we don't currently use it.  Doing so will help keep CEL consistent across the API.

Reply:
cc @jiahuif for library change.